snowshu/core/graph.py

```diff
@@ -1,28 +1,89 @@
-from typing import List, Set
-
+from typing import List, Set, Optional, Union
+from itertools import chain
 import networkx
 
 from snowshu.core.configuration_parser import Configuration
 from snowshu.core.models.relation import (Relation,
-                                          single_full_pattern_match)
+                                          single_full_pattern_match, lookup_relations)
 from snowshu.exceptions import InvalidRelationshipException
 from snowshu.logger import Logger
 
 logger = Logger().logger
 
 
 class SnowShuGraph:
-    """ Wrapper class for the networkx.Graph that represents the 
+    """ Wrapper class for the networkx.Graph that represents the
         configuration relation dependencies
 
         Attributes:
-            dag (tuple) - unused
-            graph (networkx.Graph) - Graph representation of a configuration
+            dag (Optional[tuple]) - unused
+            graph (Optional[networkx.Graph]) - Graph representation of a configuration
     """
 
     def __init__(self):
-        self.dag: tuple = None
-        self.graph: networkx.Graph = None
+        self.dag: Optional[tuple] = None
+        self.graph: Optional[networkx.Graph] = None
+
+    @staticmethod
+    def catalog_difference(source_catalog: Union["SnowShuGraph", networkx.Graph, tuple, list, set],  # noqa pylint: disable=MC0001,too-many-locals
+                           target_catalog: Union[tuple, list, set],
+                           configs: Configuration):
+        """ Builds a directed graph per replica config.
+
+            Args:
+                source_catalog: source Graph, SnowShuGraph, or catalog object which is built in current run
+                    from replica.yml file.
+                target_catalog: object which is generated by getting target catalog from existsing replica
+                    container.
+                configs: :class:`Configuration <snowshu.core.configuration_parser.Configuration>` object.
+
+            Returns:
+
+        """
+        if isinstance(source_catalog, (tuple, list, set)):
+            source = list(source_catalog)
+        elif isinstance(source_catalog, SnowShuGraph):
+            source = list(source_catalog.graph.nodes)
+        else:
+            source = list(source_catalog.nodes)
+        non_isolated_relations = []
+        for relation in configs.specified_relations:  # noqa pylint: disable=too-many-nested-blocks
```

Comment:
At the expense of comparing raw catalogs, we should use the graph's edges to determine the connections between nodes.